### DISCUSSION TITLE
Prepend to array

### DISCUSSION THREAD

**Discussion by rbraeunlich** [komapper/komapper#1325] [Q&A] [answered]
Hi everyone,
I have a Postgres table that is of type text array `text []`. Now, I want to prepend a value to this array every time a new event arrives.
Basically, I wanted to use `onDuplicateKeyUpdate().set{...}` for this but I cannot find anything that looks like array prepend. Is there any way to achieve this? Or do I have to perform a "normal" lookup and then prepend the new value?
Thanks for the help!

**Comment by nakamura-to** [comment, marked as answer]
Hello,

Here's how it can be achieved (assuming the use of JDBC in this example).
First, create a function that returns a `ColumnExpression` corresponding to PostgreSQL's `array_prepend` function.

```kotlin
private fun prepend(
    element: String,
    array: ColumnExpression<java.sql.Array, java.sql.Array>,
): ColumnExpression<java.sql.Array, java.sql.Array> {
    val name = "array_prepend"
    val o1 = Operand.SimpleArgument(typeOf<String>(), element)
    val o2 = Operand.Column(array)
    return columnExpression(name, listOf(o1, o2)) {
        append("$name(")
        append("cast(")
        visit(o1)
        append(" as text)")
        append(", ")
        visit(o2)
        append(")")
    }
}
```

Assume the entity definition is as follows.
```kotlin
@KomapperEntity
data class MyEntity(
    @KomapperId val id: Int,
    @KomapperColumn val arrayValue: Array,
)
```

The query using `onDuplicateKeyUpdate().set{...}` is as follows.
```kotlin
val m = Meta.myEntity

val arrayValue = db.config.dataFactory.createArrayOf("text", listOf("a", "b"))
val entity = MyEntity(1, arrayValue)

val query = QueryDsl.insert(m)
    .onDuplicateKeyUpdate()
    .set {
        m.arrayValue eq prepend("c", m.arrayValue)
    }
    .single(entity)
```
